File: src/charting/renderer/BarChartRenderer.java
package com.github.mikephil.charting.renderer;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;

import com.github.mikephil.charting.animation.ChartAnimator;
import com.github.mikephil.charting.buffer.BarBuffer;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.ValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.highlight.Range;
import com.github.mikephil.charting.interfaces.dataprovider.BarDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.utils.MPPointF;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.Utils;
import com.github.mikephil.charting.utils.ViewPortHandler;
import android.graphics.LinearGradient;
import com.github.mikephil.charting.model.GradientColor;

import java.util.List;

public class BarChartRenderer extends BarLineScatterCandleBubbleRenderer {

    protected BarDataProvider this.mChart;

    /**
     * the rect object that is used for drawing the bars
     */
    protected RectF this.mBarRect = new RectF();

    protected BarBuffer[] this.mBarBuffers;

    protected Paint mShadowPaint;
    protected Paint mBarBorderPaint;

    public BarChartRenderer(BarDataProvider chart, animator:ChartAnimator,
                            viewPortHandler: ViewPortHandler) {
        super(animator, viewPortHandler);
        this.mChart = chart;

        this.mHighlightPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.mHighlightPaint.setStyle(Paint.Style.FILL);
        this.mHighlightPaint.setColor(new Color(255, 0, 0, 0));
        // set alpha after color
        this.mHighlightPaint.setAlpha(120);

        this.mShadowPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.mShadowPaint.setStyle(Paint.Style.FILL);

        this.mBarBorderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.mBarBorderPaint.setStyle(Paint.Style.STROKE);
    }

    
    public initBuffers() {

        BarData barData = this.mChart.getBarData();
        this.mBarBuffers = new BarBuffer[barData.getDataSetCount()];

        for (let i = 0; i < this.mBarBuffers.length; i++) {
            IBarDataSet set = barData.getDataSetByIndex(i);
            this.mBarBuffers[i] = new BarBuffer(set.getEntryCount() * 4 * (set.isStacked() ? set.getStackSize() : 1),
                    barData.getDataSetCount(), set.isStacked());
        }
    }

    
    public drawData(c: Canvas) {

        BarData barData = this.mChart.getBarData();

        for (let i = 0; i < barData.getDataSetCount(); i++) {

            IBarDataSet set = barData.getDataSetByIndex(i);

            if (set.isVisible()) {
                drawDataSet(c, set, i);
            }
        }
    }

    private RectF this.mBarShadowRectBuffer = new RectF();

    protected drawDataSet(c: Canvas, IBarDataSet dataSet, let index) {

        Transformer trans = this.mChart.getTransformer(dataSet.getAxisDependency());

        this.mBarBorderPaint.setColor(dataSet.getBarBorderColor());
        this.mBarBorderPaint.setStrokeWidth(Utils.convertDpToPixel(dataSet.getBarBorderWidth()));

        final boolean drawBorder = dataSet.getBarBorderWidth() > 0;

        let phaseX = this.mAnimator.getPhaseX();
        let phaseY = this.mAnimator.getPhaseY();

        // draw the bar shadow before the values
        if (mChart.isDrawBarShadowEnabled()) {
            this.mShadowPaint.setColor(dataSet.getBarShadowColor());

            BarData barData = this.mChart.getBarData();

            const barWidth = barData.getBarWidth();
            const barWidthHalf = barWidth / 2.0;
            let x;

            for (let i = 0, count = Math.min((Math.ceil((dataSet.getEntryCount()) * phaseX)), dataSet.getEntryCount());
                i < count;
                i++) {

                BarEntry e = dataSet.getEntryForIndex(i);

                x = e.getX();

                this.mBarShadowRectBuffer.left = x - barWidthHalf;
                this.mBarShadowRectBuffer.right = x + barWidthHalf;

                trans.rectValueToPixel(mBarShadowRectBuffer);

                if (!mViewPortHandler.isInBoundsLeft(mBarShadowRectBuffer.right))
                    continue;

                if (!mViewPortHandler.isInBoundsRight(mBarShadowRectBuffer.left))
                    break;

                this.mBarShadowRectBuffer.top = this.mViewPortHandler.contentTop();
                this.mBarShadowRectBuffer.bottom = this.mViewPortHandler.contentBottom();

                c.drawRect(mBarShadowRectBuffer, this.mShadowPaint);
            }
        }

        // initialize the buffer
        BarBuffer buffer = this.mBarBuffers[index];
        buffer.setPhases(phaseX, phaseY);
        buffer.setDataSet(index);
        buffer.setInverted(mChart.isInverted(dataSet.getAxisDependency()));
        buffer.setBarWidth(mChart.getBarData().getBarWidth());

        buffer.feed(dataSet);

        trans.pointValuesToPixel(buffer.buffer);

        final boolean isSingleColor = dataSet.getColors().length == 1;

        if (isSingleColor) {
            this.mRenderPaint.setColor(dataSet.getColor());
        }

        for (let j = 0; j < buffer.length; j += 4) {

            if (!mViewPortHandler.isInBoundsLeft(buffer.buffer[j + 2]))
                continue;

            if (!mViewPortHandler.isInBoundsRight(buffer.buffer[j]))
                break;

            if (!isSingleColor) {
                // Set the color for the currently drawn value. If the index
                // is out of bounds, reuse colors.
                this.mRenderPaint.setColor(dataSet.getColor(j / 4));
            }

            if (dataSet.getGradientColor() != null) {
                GradientColor gradientColor = dataSet.getGradientColor();
                 this.mRenderPaint.setShader(
                    new LinearGradient(
                        buffer.buffer[j],
                        buffer.buffer[j + 3],
                        buffer.buffer[j],
                        buffer.buffer[j + 1],
                        gradientColor.getStartColor(),
                        gradientColor.getEndColor(),
                        android.graphics.Shader.TileMode.MIRROR));
            }

            if (dataSet.getGradientColors() != null) {
                 this.mRenderPaint.setShader(
                    new LinearGradient(
                        buffer.buffer[j],
                        buffer.buffer[j + 3],
                        buffer.buffer[j],
                        buffer.buffer[j + 1],
                        dataSet.getGradientColor(j / 4).getStartColor(),
                        dataSet.getGradientColor(j / 4).getEndColor(),
                        android.graphics.Shader.TileMode.MIRROR));
            }


            c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                    buffer.buffer[j + 3], this.mRenderPaint);

            if (drawBorder) {
                c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                        buffer.buffer[j + 3], this.mBarBorderPaint);
            }
        }
    }

    protected prepareBarHighlight(let x, let y1, let y2, let barWidthHalf, Transformer trans) {

        let left = x - barWidthHalf;
        let right = x + barWidthHalf;
        let top = y1;
        let bottom = y2;

        this.mBarRect.set(left, top, right, bottom);

        trans.rectToPixelPhase(mBarRect, this.mAnimator.getPhaseY());
    }

    
    public drawValues(c: Canvas) {

        // if values are drawn
        if (isDrawingValuesAllowed(mChart)) {

            List<IBarDataSet> dataSets = this.mChart.getBarData().getDataSets();

            const valueOffsetPlus = Utils.convertDpToPixel(4.5f);
            let posOffset = 0;
            let negOffset = 0;
            boolean drawValueAboveBar = this.mChart.isDrawValueAboveBarEnabled();

            for (let i = 0; i < this.mChart.getBarData().getDataSetCount(); i++) {

                IBarDataSet dataSet = dataSets.get(i);

                if (!shouldDrawValues(dataSet))
                    continue;

                // apply the text-styling defined by the DataSet
                applyValueTextStyle(dataSet);

                boolean isInverted = this.mChart.isInverted(dataSet.getAxisDependency());

                // calculate the correct offset depending on the draw position of
                // the value
                let valueTextHeight = Utils.calcTextHeight(mValuePaint, "8");
                posOffset = (drawValueAboveBar ? -valueOffsetPlus : valueTextHeight + valueOffsetPlus);
                negOffset = (drawValueAboveBar ? valueTextHeight + valueOffsetPlus : -valueOffsetPlus);

                if (isInverted) {
                    posOffset = -posOffset - valueTextHeight;
                    negOffset = -negOffset - valueTextHeight;
                }

                // get the buffer
                BarBuffer buffer = this.mBarBuffers[i];

                const phaseY = this.mAnimator.getPhaseY();

                ValueFormatter formatter = dataSet.getValueFormatter();

                MPPointF iconsOffset = MPPointF.getInstance(dataSet.getIconsOffset());
                iconsOffset.x = Utils.convertDpToPixel(iconsOffset.x);
                iconsOffset.y = Utils.convertDpToPixel(iconsOffset.y);

                // if only single values are drawn (sum)
                if (!dataSet.isStacked()) {

                    for (let j = 0; j < buffer.buffer.length * this.mAnimator.getPhaseX(); j += 4) {

                        let x = (buffer.buffer[j] + buffer.buffer[j + 2]) / 2f;

                        if (!mViewPortHandler.isInBoundsRight(x))
                            break;

                        if (!mViewPortHandler.isInBoundsY(buffer.buffer[j + 1])
                                || !mViewPortHandler.isInBoundsLeft(x))
                            continue;

                        BarEntry entry = dataSet.getEntryForIndex(j / 4);
                        let val = entry.getY();

                        if (dataSet.isDrawValuesEnabled()) {
                            drawValue(c, formatter.getBarLabel(entry), x, val >= 0 ?
                                            (buffer.buffer[j + 1] + posOffset) :
                                            (buffer.buffer[j + 3] + negOffset),
                                    dataSet.getValueTextColor(j / 4));
                        }

                        if (entry.getIcon() != null && dataSet.isDrawIconsEnabled()) {

                            Drawable icon = entry.getIcon();

                            let px = x;
                            let py = val >= 0 ?
                                    (buffer.buffer[j + 1] + posOffset) :
                                    (buffer.buffer[j + 3] + negOffset);

                            px += iconsOffset.x;
                            py += iconsOffset.y;

                            Utils.drawImage(
                                    c,
                                    icon,
                                    px,
                                    py,
                                    icon.getIntrinsicWidth(),
                                    icon.getIntrinsicHeight());
                        }
                    }

                    // if we have stacks
                } else {

                    Transformer trans = this.mChart.getTransformer(dataSet.getAxisDependency());

                    let bufferIndex = 0;
                    let index = 0;

                    while (index < dataSet.getEntryCount() * this.mAnimator.getPhaseX()) {

                        BarEntry entry = dataSet.getEntryForIndex(index);

                        float[] vals = entry.getYVals();
                        let x = (buffer.buffer[bufferIndex] + buffer.buffer[bufferIndex + 2]) / 2f;

                        let color = dataSet.getValueTextColor(index);

                        // we still draw stacked bars, but there is one
                        // non-stacked
                        // in between
                        if (vals == null) {

                            if (!mViewPortHandler.isInBoundsRight(x))
                                break;

                            if (!mViewPortHandler.isInBoundsY(buffer.buffer[bufferIndex + 1])
                                    || !mViewPortHandler.isInBoundsLeft(x))
                                continue;

                            if (dataSet.isDrawValuesEnabled()) {
                                drawValue(c, formatter.getBarLabel(entry), x, buffer.buffer[bufferIndex + 1] +
                                                (entry.getY() >= 0 ? posOffset : negOffset),
                                        color);
                            }

                            if (entry.getIcon() != null && dataSet.isDrawIconsEnabled()) {

                                Drawable icon = entry.getIcon();

                                let px = x;
                                let py = buffer.buffer[bufferIndex + 1] +
                                        (entry.getY() >= 0 ? posOffset : negOffset);

                                px += iconsOffset.x;
                                py += iconsOffset.y;

                                Utils.drawImage(
                                        c,
                                        icon,
                                        px,
                                        py,
                                        icon.getIntrinsicWidth(),
                                        icon.getIntrinsicHeight());
                            }

                            // draw stack values
                        } else {

                            float[] transformed = new float[vals.length * 2];

                            let posY = 0;
                            let negY = -entry.getNegativeSum();

                            for (let k = 0, idx = 0; k < transformed.length; k += 2, idx++) {

                                let value = vals[idx];
                                let y;

                                if (value == 0.0 && (posY == 0.0 || negY == 0.0)) {
                                    // Take care of the situation of a 0.0 value, which overlaps a non-zero bar
                                    y = value;
                                } else if (value >= 0.0) {
                                    posY += value;
                                    y = posY;
                                } else {
                                    y = negY;
                                    negY -= value;
                                }

                                transformed[k + 1] = y * phaseY;
                            }

                            trans.pointValuesToPixel(transformed);

                            for (let k = 0; k < transformed.length; k += 2) {

                                const val = vals[k / 2];
                                final boolean drawBelow =
                                        (val == 0.0 && negY == 0.0 && posY > 0.0) ||
                                                val < 0.0;
                                let y = transformed[k + 1]
                                        + (drawBelow ? negOffset : posOffset);

                                if (!mViewPortHandler.isInBoundsRight(x))
                                    break;

                                if (!mViewPortHandler.isInBoundsY(y)
                                        || !mViewPortHandler.isInBoundsLeft(x))
                                    continue;

                                if (dataSet.isDrawValuesEnabled()) {
                                    drawValue(c, formatter.getBarStackedLabel(val, entry), x, y, color);
                                }

                                if (entry.getIcon() != null && dataSet.isDrawIconsEnabled()) {

                                    Drawable icon = entry.getIcon();

                                    Utils.drawImage(
                                            c,
                                            icon,
                                            (x + iconsOffset.x),
                                            (y + iconsOffset.y),
                                            icon.getIntrinsicWidth(),
                                            icon.getIntrinsicHeight());
                                }
                            }
                        }

                        bufferIndex = vals == null ? bufferIndex + 4 : bufferIndex + 4 * vals.length;
                        index++;
                    }
                }

                MPPointF.recycleInstance(iconsOffset);
            }
        }
    }

    
    public drawValue(c: Canvas, let valueText, let x, let y, let color) {
        this.mValuePaint.setColor(color);
        c.drawText(valueText, x, y, this.mValuePaint);
    }

    
    public drawHighlighted(c: Canvas, Highlight[] indices) {

        BarData barData = this.mChart.getBarData();

        for (Highlight high : indices) {

            IBarDataSet set = barData.getDataSetByIndex(high.getDataSetIndex());

            if (set == null || !set.isHighlightEnabled())
                continue;

            BarEntry e = set.getEntryForXValue(high.getX(), high.getY());

            if (!isInBoundsX(e, set))
                continue;

            Transformer trans = this.mChart.getTransformer(set.getAxisDependency());

            this.mHighlightPaint.setColor(set.getHighLightColor());
            this.mHighlightPaint.setAlpha(set.getHighLightAlpha());

            boolean isStack = (high.getStackIndex() >= 0  && e.isStacked()) ? true : false;

            const y1;
            const y2;

            if (isStack) {

                if(mChart.isHighlightFullBarEnabled()) {

                    y1 = e.getPositiveSum();
                    y2 = -e.getNegativeSum();

                } else {

                    Range range = e.getRanges()[high.getStackIndex()];

                    y1 = range.from;
                    y2 = range.to;
                }

            } else {
                y1 = e.getY();
                y2 = 0;
            }

            prepareBarHighlight(e.getX(), y1, y2, barData.getBarWidth() / 2f, trans);

            setHighlightDrawPos(high, this.mBarRect);

            c.drawRect(mBarRect, this.mHighlightPaint);
        }
    }

    /**
     * Sets the drawing position of the highlight object based on the riven bar-rect.
     * @param high
     */
    protected setHighlightDrawPos(Highlight high, RectF bar) {
        high.setDraw(bar.centerX(), bar.top);
    }

    
    public drawExtras(c: Canvas) {
    }
}

File: src/charting/renderer/BarChartRenderer.ts
import { BarLineScatterCandleBubbleRenderer } from './BarLineScatterCandleBubbleRenderer';
import { ChartAnimator } from '../animation/ChartAnimator';
import { BarBuffer } from '../buffer/BarBuffer';
import { BarChart, CustomRenderer } from '../charts/BarChart';
import { Highlight } from '../highlight/Highlight';
import { IBarDataSet } from '../interfaces/datasets/IBarDataSet';
import { Transformer } from '../utils/Transformer';
import { Utils } from '../utils/Utils';
import { ViewPortHandler } from '../utils/ViewPortHandler';
import { Canvas, Paint, RectF, Style } from '@nativescript-community/ui-canvas';
import { profile } from '@nativescript/core';
import { Entry } from '../data/Entry';

export class BarChartRenderer extends BarLineScatterCandleBubbleRenderer {
    public mChart: BarChart;

    protected mBarBuffers: BarBuffer[];

    /**
     * palet for the bar shadow
     */
    protected mShadowPaint: Paint;

    /**
     * palet for the bar border
     */
    protected mBarBorderPaint: Paint;

    protected mBarRect = new RectF(0, 0, 0, 0);
    protected get barRect() {
        if (!this.mBarRect) {
            this.mBarRect = new RectF(0, 0, 0, 0);
        }
        return this.mBarRect;
    }
    protected mBarShadowRectBuffer: RectF;
    protected get barShadowRectBuffer() {
        if (!this.mBarShadowRectBuffer) {
            this.mBarShadowRectBuffer = new RectF(0, 0, 0, 0);
        }
        return this.mBarShadowRectBuffer;
    }

    constructor(chart: BarChart, animator: ChartAnimator, viewPortHandler: ViewPortHandler) {
        super(animator, viewPortHandler);
        this.mChart = chart;
    }

    public get highlightPaint() {
        if (!this.mHighlightPaint) {
            this.mHighlightPaint = new Paint();
            this.mHighlightPaint.setAntiAlias(true);
            this.mHighlightPaint.setStyle(Style.FILL);
            this.mHighlightPaint.setColor('black');
            // set alpha after color
            this.mHighlightPaint.setAlpha(120);
        }
        return this.mHighlightPaint;
    }
    public get barBorderPaint() {
        if (!this.mBarBorderPaint) {
            this.mBarBorderPaint = new Paint();
            this.mBarBorderPaint.setAntiAlias(true);
            this.mBarBorderPaint.setStyle(Style.STROKE);
        }
        return this.mBarBorderPaint;
    }
    public get shadowPaint() {
        if (!this.mShadowPaint) {
            this.mShadowPaint = new Paint();
            this.mShadowPaint.setAntiAlias(true);
            this.mShadowPaint.setStyle(Style.FILL);
        }
        return this.mShadowPaint;
    }

    public initBuffers() {
        const barData = this.mChart.getBarData();
        this.mBarBuffers = [];

        for (let i = 0; i < barData.getDataSetCount(); i++) {
            const set = barData.getDataSetByIndex(i);
            this.mBarBuffers.push(new BarBuffer(set.getEntryCount() * 4 * (set.isStacked() ? set.getStackSize() : 1), barData.getDataSetCount(), set.isStacked()));
        }
    }

    @profile
    public drawData(c: Canvas) {
        const barData = this.mChart.getBarData();

        for (let i = 0; i < barData.getDataSetCount(); i++) {
            const set = barData.getDataSetByIndex(i);
            if (set.isVisible()) {
                this.drawDataSet(c, set, i);
            }
        }
    }

    protected drawDataSet(c: Canvas, dataSet: IBarDataSet, index: number): boolean {
        const trans = this.mChart.getTransformer(dataSet.getAxisDependency());

        const drawBorder = dataSet.getBarBorderWidth() > 0;
        let borderPaint: Paint;
        if (drawBorder) {
            borderPaint = this.barBorderPaint;
            borderPaint.setColor(dataSet.getBarBorderColor());
            borderPaint.setStrokeWidth(dataSet.getBarBorderWidth());
        }

        const phaseX = this.mAnimator.getPhaseX();
        const phaseY = this.mAnimator.getPhaseY();

        // draw the bar shadow before the values
        if (this.mChart.isDrawBarShadowEnabled()) {
            const paint = this.shadowPaint;
            paint.setColor(dataSet.getBarShadowColor());

            const barData = this.mChart.getBarData();

            const barWidth = barData.getBarWidth();
            const barWidthHalf = barWidth / 2;
            let x;

            const barShadowRectBuffer = this.barShadowRectBuffer;
            for (let i = 0, count = Math.min(Math.ceil(dataSet.getEntryCount() * phaseX), dataSet.getEntryCount()); i < count; i++) {
                const e = dataSet.getEntryForIndex(i);
                x = dataSet.getEntryXValue(e, i);
                barShadowRectBuffer.left = x - barWidthHalf;
                barShadowRectBuffer.right = x + barWidthHalf;

                trans.rectValueToPixel(barShadowRectBuffer);

                if (!this.mViewPortHandler.isInBoundsLeft(barShadowRectBuffer.right)) {
                    continue;
                }

                if (!this.mViewPortHandler.isInBoundsRight(barShadowRectBuffer.left)) {
                    break;
                }

                barShadowRectBuffer.top = this.mViewPortHandler.contentTop();
                barShadowRectBuffer.bottom = this.mViewPortHandler.contentBottom();

                c.drawRect(barShadowRectBuffer, paint);
            }
        }

        // initialize the buffer
        const buffer = this.mBarBuffers[index];
        buffer.setPhases(phaseX, phaseY);
        buffer.setDataSet(index);
        buffer.setInverted(this.mChart.isInverted(dataSet.getAxisDependency()));
        buffer.setBarWidth(this.mChart.getBarData().getBarWidth());
        buffer.setYAxisMin(this.mChart.getAxis(dataSet.getAxisDependency()).getAxisMinimum());
        buffer.setYAxisMax(this.mChart.getAxis(dataSet.getAxisDependency()).getAxisMaximum());

        buffer.feed(dataSet);

        trans.pointValuesToPixel(buffer.buffer);

        const isSingleColor = dataSet.getColors().length === 1;
        // const isInverted = this.mChart.isInverted(dataSet.getAxisDependency());
        const renderPaint = this.renderPaint;
        const previousShader = renderPaint.getShader();
        const shader = dataSet.getFillShader();
        if (shader) {
            renderPaint.setShader(shader);
        }
        if (isSingleColor) {
            renderPaint.setColor(dataSet.getColor());
        }

        const customRender = this.mChart.getCustomRenderer();
        for (let j = 0; j < buffer.size(); j += 4) {
            if (!this.mViewPortHandler.isInBoundsLeft(buffer.buffer[j + 2])) {
                continue;
            }

            if (!this.mViewPortHandler.isInBoundsRight(buffer.buffer[j])) {
                break;
            }
            if (!isSingleColor) {
                // Set the color for the currently drawn value. If the index
                // is out of bounds, reuse colors.
                renderPaint.setColor(dataSet.getColor(j / 4));
            }
            if (customRender && customRender.drawBar) {
                const e = dataSet.getEntryForIndex(j / 4);
                customRender.drawBar(c, e, dataSet, buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2], buffer.buffer[j + 3], renderPaint);
            } else {
                c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2], buffer.buffer[j + 3], renderPaint);

                if (drawBorder) {
                    c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2], buffer.buffer[j + 3], borderPaint);
                }
            }
        }
        renderPaint.setShader(previousShader);

        return true;
    }

    protected prepareBarHighlight(x: number, y1: number, y2: number, barWidthHalf: number, trans: Transformer) {
        const left = x - barWidthHalf;
        const right = x + barWidthHalf;
        const top = y1;
        const bottom = y2;
        const barRect = this.barRect;
        barRect.set(left, top, right, bottom);

        trans.rectToPixelPhase(barRect, this.mAnimator.getPhaseY());
    }

    public drawValues(c: Canvas) {
        const data = this.mChart.getData();
        const dataSets = data.getDataSets();
        if (!this.isDrawingValuesAllowed(this.mChart) || dataSets.some((d) => d.isDrawValuesEnabled() || d.isDrawIconsEnabled()) === false) {
            return;
        }
        // if values are drawn

        const valueOffsetPlus = 4.5;
        let posOffset = 0;
        let negOffset = 0;
        const drawValueAboveBar = this.mChart.isDrawValueAboveBarEnabled();
        const paint = this.valuePaint;
        const customRender = this.mChart.getCustomRenderer();
        for (let i = 0; i < this.mChart.getBarData().getDataSetCount(); i++) {
            const dataSet = dataSets[i];
            if (!this.shouldDrawValues(dataSet)) {
                continue;
            }

            const yKey = dataSet.yProperty;

            // apply the text-styling defined by the DataSet
            this.applyValueTextStyle(dataSet);

            const isInverted = this.mChart.isInverted(dataSet.getAxisDependency());

            // calculate the correct offset depending on the draw position of
            // the value
            const valueTextHeight = Utils.calcTextHeight(paint, '8');
            const valuesOffset = dataSet.getValuesOffset();
            posOffset = drawValueAboveBar ? -(valueOffsetPlus + valuesOffset.y) : valueTextHeight + (valueOffsetPlus + valuesOffset.y);
            negOffset = drawValueAboveBar ? valueTextHeight + (valueOffsetPlus + valuesOffset.y) : -(valueOffsetPlus + valuesOffset.y);

            if (isInverted) {
                posOffset = -posOffset - valueTextHeight;
                negOffset = -negOffset - valueTextHeight;
            }

            // get the buffer
            const buffer = this.mBarBuffers[i];

            const phaseY = this.mAnimator.getPhaseY();
            const formatter = dataSet.getValueFormatter();

            const iconsOffset = dataSet.getIconsOffset();

            // if only single values are drawn (sum)

            const isDrawValuesEnabled = dataSet.isDrawValuesEnabled();
            const isDrawIconsEnabled = dataSet.isDrawIconsEnabled();
            if (!dataSet.isStacked()) {
                for (let j = 0; j < buffer.size() * this.mAnimator.getPhaseX(); j += 4) {
                    const x = (buffer.buffer[j] + buffer.buffer[j + 2]) / 2;

                    if (!this.mViewPortHandler.isInBoundsRight(x)) {
                        break;
                    }

                    const entry = dataSet.getEntryForIndex(j / 4);
                    const val = entry[yKey];

                    if (!this.mViewPortHandler.isInBoundsY(buffer.buffer[j + (val >= 0 ? 1 : 3)]) || !this.mViewPortHandler.isInBoundsLeft(x)) {
                        continue;
                    }

                    if (isDrawValuesEnabled) {
                        this.drawValue(
                            c,
                            formatter.getBarLabel(val, entry),
                            x + valuesOffset.x,
                            val >= 0 ? buffer.buffer[j + 1] + posOffset : buffer.buffer[j + 3] + negOffset,
                            dataSet.getValueTextColor(j / 4),
                            paint,
                            customRender
                        );
                    }

                    if (isDrawIconsEnabled && entry.icon != null) {
                        const icon = entry.icon;

                        let px = x;
                        let py = val >= 0 ? buffer.buffer[j + 1] + posOffset : buffer.buffer[j + 3] + negOffset;

                        px += iconsOffset.x;
                        py += iconsOffset.y;

                        Utils.drawIcon(c, this.mChart, icon, px, py);
                    }
                }
                // if we have stacks
            } else {
                const trans = this.mChart.getTransformer(dataSet.getAxisDependency());

                let bufferIndex = 0;
                let index = 0;

                while (index < dataSet.getEntryCount() * this.mAnimator.getPhaseX()) {
                    const entry = dataSet.getEntryForIndex(index);

                    const vals = entry.yVals;
                    const x = (buffer.buffer[bufferIndex] + buffer.buffer[bufferIndex + 2]) / 2;

                    const color = dataSet.getValueTextColor(index);

                    // we still draw stacked bars, but there is one
                    // non-stacked
                    // in between
                    if (vals == null) {
                        if (!this.mViewPortHandler.isInBoundsRight(x)) {
                            break;
                        }

                        if (!this.mViewPortHandler.isInBoundsY(buffer.buffer[bufferIndex + (entry[yKey] >= 0 ? 1 : 3)]) || !this.mViewPortHandler.isInBoundsLeft(x)) {
                            continue;
                        }

                        if (isDrawValuesEnabled) {
                            this.drawValue(
                                c,
                                formatter.getBarLabel(entry[yKey], entry),
                                x,
                                entry[yKey] >= 0 ? buffer.buffer[bufferIndex + 1] + posOffset : buffer.buffer[bufferIndex + 3] + negOffset,
                                color,
                                paint,
                                customRender
                            );
                        }

                        if (isDrawIconsEnabled && entry.icon != null) {
                            const icon = entry.icon;

                            let px = x;
                            let py = entry[yKey] >= 0 ? buffer.buffer[bufferIndex + 1] + posOffset : buffer.buffer[bufferIndex + 3] + negOffset;

                            px += iconsOffset.x;
                            py += iconsOffset.y;

                            Utils.drawIcon(c, this.mChart, icon, px, py);
                        }
                        // draw stack values
                    } else {
                        const transformed = Utils.createNativeArray(vals.length * 2);

                        let posY = 0;
                        let negY = -entry.negativeSum;

                        for (let k = 0, idx = 0; k < transformed.length; k += 2, idx++) {
                            const value = vals[idx];
                            let y;

                            if (value === 0 && (posY === 0 || negY === 0)) {
                                // Take care of the situation of a 0.0 value, which overlaps a non-zero bar
                                y = value;
                            } else if (value >= 0) {
                                posY += value;
                                y = posY;
                            } else {
                                y = negY;
                                negY -= value;
                            }

                            transformed[k + 1] = y * phaseY;
                        }

                        trans.pointValuesToPixel(transformed);

                        for (let k = 0; k < transformed.length; k += 2) {
                            const val = vals[k / 2];
                            const drawBelow = (val === 0 && negY === 0 && posY > 0) || val < 0;
                            const y = transformed[k + 1] + (drawBelow ? negOffset : posOffset);

                            if (!this.mViewPortHandler.isInBoundsRight(x)) {
                                break;
                            }

                            if (!this.mViewPortHandler.isInBoundsY(y) || !this.mViewPortHandler.isInBoundsLeft(x)) {
                                continue;
                            }

                            if (dataSet.isDrawValuesEnabled()) {
                                this.drawValue(c, formatter.getBarStackedLabel(val, entry), x, y, color, paint, customRender);
                            }

                            if (entry.icon != null && dataSet.isDrawIconsEnabled()) {
                                const icon = entry.icon;
                                Utils.drawIcon(c, this.mChart, icon, x + iconsOffset.x, y + iconsOffset.y);
                            }
                        }
                    }

                    bufferIndex = vals == null ? bufferIndex + 4 : bufferIndex + 4 * vals.length;
                    index++;
                }
            }
        }
    }

    public drawHighlighted(c: Canvas, indices: Highlight[]) {
        const barData = this.mChart.getBarData();
        let entry: Entry, index: number;
        const barRect = this.barRect;
        for (let i = 0; i < indices.length; i++) {
            const high = indices[i];
            const set = barData.getDataSetByIndex(high.dataSetIndex);

            if (set === null || !set.isHighlightEnabled()) {
                continue;
            }
            if (high.entry) {
                entry = high.entry;
                index = high.entryIndex;
            } else {
                const r = set.getEntryAndIndexForXValue(high.x, high.y);
                entry = r.entry;
                index = r.index;
            }
            if (!this.isInBoundsX(entry, set)) {
                continue;
            }

            const yKey = set.yProperty;
            const trans = this.mChart.getTransformer(set.getAxisDependency());

            const paint = this.highlightPaint;
            paint.setColor(set.getHighLightColor());
            paint.setAlpha(set.getHighLightAlpha());

            const isStack = high.stackIndex >= 0 && entry.isStacked ? true : false;

            let y1;
            let y2;

            if (isStack) {
                if (this.mChart.isHighlightFullBarEnabled()) {
                    y1 = entry.positiveSum;
                    y2 = -entry.negativeSum;
                } else {
                    const range = entry.ranges[high.stackIndex];
                    y1 = range[0];
                    y2 = range[1];
                }
            } else {
                const minAxisValue = this.mChart.getAxis(set.getAxisDependency()).getAxisMinimum();
                y1 = entry[yKey];
                y2 = minAxisValue >= 0 ? minAxisValue : 0;
            }
            const x = set.getEntryXValue(entry, index);
            this.prepareBarHighlight(x, y1, y2, barData.getBarWidth() / 2, trans);

            this.setHighlightDrawPos(high, barRect);
            const customRender = this.mChart.getCustomRenderer();
            if (customRender && customRender.drawHighlight) {
                const rect = barRect;
                customRender.drawHighlight(c, high, rect.left, rect.top, rect.right, rect.bottom, paint);
            } else {
                c.drawRect(barRect, paint);
            }
        }
    }

    /**
     * Sets the drawing position of the highlight object based on the riven bar-rect.
     * @param high
     * @param bar
     */
    protected setHighlightDrawPos(high: Highlight, bar: RectF) {
        high.drawX = bar.centerX();
        high.drawY = bar.top;
    }

    public drawExtras(c: Canvas) {}
}
